### PR TITLE
4.1 formhelper value sources

### PR DESCRIPTION
**Pull request by ADmad**
- Throw exception for invalid source values instead of silently ignoring them.
- Improved docblocks

// @diegosardina

### CODE REVIEW

File: src/View/Helper/FormHelper.php [hunk 1]
```diff
@@ -2488,28 +2494,41 @@ public function getValueSources(): array
...
      *
-     * Valid values are set in `$supportedValueSources`.
-     * You need to supply one valid context or multiple, as a list of strings. Order sets priority.
+     * You need to supply one or more valid source, as a list of strings.
+     * Order sets priority.
      *
+     * @see FormHelper::$supportedValueSources for valid values.
      * @param string|string[] $sources A string or a list of strings identifying a source.
      * @return $this
      */
```
Comment (othercorey) [inline src/View/Helper/FormHelper.php:2525]:
```suggestion
     * @throws \InvalidArgumentException If sources list contains invalid value.
     */
```

Isn't this the public api for users?

File: src/View/Helper/FormHelper.php [hunk 2]
```diff
@@ -2488,28 +2494,42 @@ public function getValueSources(): array
...
-     * Extracts valid value sources.
+     * Validate value sources.
      *
      * @param string[] $sources A list of strings identifying a source.
-     * @return string[]
+     * @return void
+     * @throws \InvalidArgumentException If sources list contains invalid value.
      */
-    protected function extractValidValueSources(array $sources): array
+    protected function validateValueSources(array $sources): void
```
Comment (saeideng) [inline src/View/Helper/FormHelper.php:2503]: removing old `extractValidValueSources` methods is not BC !! is ?

Comment (diegosardina) [reply src/View/Helper/FormHelper.php:2503]: No, it was added by me some days ago because sourceValues array was hardcoded in FormHelper

Comment (dereuromark) [reply src/View/Helper/FormHelper.php:2503]: If it is a protected method, then in a minor this wouldn't be too big of a deal. At least if it is very unlikely to be overwritten - which it seems in this case.

Comment (ADmad) [reply src/View/Helper/FormHelper.php:2503]: > removing old extractValidValueSources methods is not BC !! is ?

No it's not. This method is not present in any release. It was added just few days ago to 4.next branch.

Even if it was present in 4.0, our BC policy allows changing protected methods in new minors.

File: src/View/Helper/FormHelper.php [hunk 3]
```diff
@@ -2488,28 +2494,42 @@ public function getValueSources(): array
...
+            ));
+        }
     }
 
     /**
      * Sets the value sources.
      *
-     * Valid values are set in `$supportedValueSources`.
-     * You need to supply one valid context or multiple, as a list of strings. Order sets priority.
+     * You need to supply one or more valid source, as a list of strings.
```
Comment (markstory) [inline src/View/Helper/FormHelper.php:2519]:
```suggestion
     * You need to supply one or more valid sources, as a list of strings.
```

**Pull request merged**